cache/compressed_secondary_cache.cc

```diff
@@ -145,6 +139,96 @@ std::string CompressedSecondaryCache::GetPrintableOptions() const {
   return ret;
 }
 
+CacheAllocationPtr CompressedSecondaryCache::SplitValueIntoChunks(
+    const Slice& value, const CompressionType compression_type,
+    size_t& charge) {
+  assert(!value.empty());
+  const char* src_ptr = value.data();
+  size_t src_size{value.size()};
+
+  CacheValueChunk* dummy_head = new CacheValueChunk();
+  CacheValueChunk* current_chunk = dummy_head;
+  size_t predicted_chunk_size{0};
+  size_t actual_chunk_size{0};
+  size_t tmp_size{0};
+  while (src_size > 0) {
+    predicted_chunk_size = sizeof(CacheValueChunk) - 1 + src_size;
+    auto upper =
+        std::upper_bound(malloc_bin_sizes_.begin(), malloc_bin_sizes_.end(),
+                         predicted_chunk_size);
+    // Do not split when value size is too small, too large, or there is no
+    // compression.
+    if (upper == malloc_bin_sizes_.begin() ||
+        upper == malloc_bin_sizes_.end() ||
+        *upper - predicted_chunk_size < malloc_bin_sizes_.front() ||
+        compression_type == kNoCompression) {
+      tmp_size = predicted_chunk_size;
+    } else {
+      tmp_size = *(--upper);
+    }
+
+    CacheAllocationPtr next_ptr =
+        AllocateBlock(tmp_size, memory_allocator_.get());
+    CacheValueChunk* next_chunk =
+        reinterpret_cast<CacheValueChunk*>(next_ptr.get());
+    actual_chunk_size = tmp_size - sizeof(CacheValueChunk) + 1;
+    memcpy(next_chunk->data, src_ptr, actual_chunk_size);
+    next_chunk->size = actual_chunk_size;
+    current_chunk->next = std::move(next_ptr);
+    current_chunk = next_chunk;
+    src_ptr += actual_chunk_size;
+    src_size -= actual_chunk_size;
+    charge += tmp_size;
+  }
+  // Last chunk's next is dummy.
+  CacheAllocationPtr dummy_ptr =
+      AllocateBlock(sizeof(CacheValueChunk), memory_allocator_.get());
+  current_chunk->next = std::move(dummy_ptr);
```

Comment:
USE_CLANG=1 COMPILE_WITH_ASAN=1  make -j 24

$ ./compressed_secondary_cache_test --gtest_filter="*SplitValueIntoChunksTest" 

failure log: https://www.internalfb.com/phabricator/paste/view/P519678071


backtace in https://www.internalfb.com/phabricator/paste/view/P519677799